feat: allow customization of taints for node drain detection

MarcusDunn opened the pull request:
Introduce the `DRAIN_TAINTS` operator configuration option, enabling users to specify which taint keys should be recognized as indicators of node drain.

By default, this includes commonly used taints from `kubectl`, Cluster Autoscaler, and Karpenter:

- `node.kubernetes.io/unschedulable`
- `ToBeDeletedByClusterAutoscaler`
- `karpenter.sh/disrupted`
- `karpenter.sh/disruption`

This enhancement provides greater flexibility in handling node drain scenarios, allowing adaptation to different cluster configurations and scheduling strategies.

Fixes: #5299

MarcusDunn committed:
feat: check for karpenter taint when looking for cordoned nodes

Signed-off-by: Marcus Dunn <marcus.s.dunn@gmail.com>

MarcusDunn committed:
fix: remove taint from node after test

Signed-off-by: Marcus Dunn <marcus.s.dunn@gmail.com>

MarcusDunn committed:
feat: allow configuring drain taints

Signed-off-by: Marcus Dunn <marcus.s.dunn@gmail.com>

FF-Ninja commented on internal/controller/cluster_controller.go:
I think you need to keep this logic but change to isNodeBeingDrained. Without it, every node update will trigger a complete reconciliation loop, which will get expensive, especially on large clusters that have node churn.

MarcusDunn commented:
@sxd could you take a look at this one? apologies for the three separate PRs this seemed to turn into.

armru commented:
/test limit=local depth=push tl=4

MarcusDunn committed:
fix: updated api docs

Signed-off-by: Marcus Dunn <marcus.s.dunn@gmail.com>

MarcusDunn committed:
fix: fixed failing lints

Mostly via formatting but had to ignore one because I did not want to spread URL across two lines.

Signed-off-by: Marcus Dunn <marcus.s.dunn@gmail.com>

FF-Ninja commented:
Hi @MarcusDunn 

Hope you’re well. 

I think you may need to add back the drain taints check to the cluster_controller.go. Without that, I think it will trigger NodeToCluster reconciliation loop whenever a node is changed, added, removed from the cluster and it should only do that if a node is draining I believe. 

Thanks

MarcusDunn commented:
I would have liked to have a more clever check in `cluster_controller` but because the drain taints are configured per-cluster, I don't think there was a great way to do it.

For now (I think!) it only runs if any taints have changed.

Suggestions welcome.

MarcusDunn commented:
That seems pretty reasonable.

MarcusDunn committed:
feat: moved configuration from cluster to the operator and tightened up when reconcile is triggered

Signed-off-by: Marcus Dunn <marcus.s.dunn@gmail.com>

MarcusDunn committed:
docs: added docs for `DRAIN_TAINTS`

Signed-off-by: Marcus Dunn <marcus.s.dunn@gmail.com>

MarcusDunn committed:
docs: removed old docs

Signed-off-by: Marcus Dunn <marcus.s.dunn@gmail.com>

armru commented:
/test limit=local depth=push tl=4

MarcusDunn commented:
@armru Is there anything I can do to move this along?

armru commented:
@mnencia we should put this in the queue so we can take a look and give feedback

leonardoce committed:
chore: review, nits, unit tests

Signed-off-by: Leonardo Cecchi <leonardo.cecchi@enterprisedb.com>

leonardoce commented:
@MarcusDunn, thank you for this contribution. It was really good!
I just changed a few nits and added some unit tests. Do you agree with what I changed?

leonardoce commented:
/test limit=local depth=push tl=4

MarcusDunn commented:
> @MarcusDunn, thank you for this contribution. It was really good! I just changed a few nits and added some unit tests. Do you agree with what I changed?

Changes look great, thanks for taking a look! 

just for my own sake, when are struct fields PaselCase vs camelCase? I noticed you swapped `DrainTaints` to `drainTaints`

leonardoce commented:
> just for my own sake, when are struct fields PaselCase vs camelCase? I noticed you swapped DrainTaints to drainTaints. This is my first attempt at writing any go, so it's all pretty new.

`PaselCase` fields [are exported while](https://go.dev/ref/spec#Exported_identifiers) `camelCase` fields are not.
Yeah, it sounds weird if you are not used to it.
Your patch looked pretty good, and it is surprising that this is your first attempt. 
That means you've a strong background. Congrats!

armru committed:
test: better descriptions

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

mnencia committed:
chore: licenses

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

armru committed:
test: revert changes

Signed-off-by: Armando Ruocco <armando.ruocco@enterprisedb.com>

armru commented:
/test limit=local tl=4 depth=push

mnencia committed:
chore: typo fix

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
docs: improve

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>
